lib/src/document.dart

```diff
@@ -97,6 +106,81 @@ class Document {
       }
     }
   }
+
+  /// Footnotes could be defined in arbitrary positions of a document, we need
+  /// to distinguish them and put them behind; and every footnote definition
+  /// may have multiple backrefs, we need to append backrefs for it.
+  List<Node> _filterFootnotes(List<Node> nodes) {
+    final footnotes = <Element>[];
+    final blocks = <Node>[];
+    for (final node in nodes) {
+      if (node is Element &&
+          node.tag == 'li' &&
+          footnoteReferences.containsKey(node.footnoteLabel)) {
+        final label = node.footnoteLabel;
+        var count = 0;
+        if (label != null && (count = footnoteReferences[label] ?? 0) > 0) {
+          footnotes.add(node);
+          final children = node.children;
+          if (children != null) {
+            _appendBackref(children, Uri.encodeComponent(label), count);
+          }
+        }
+      } else {
+        blocks.add(node);
+      }
+    }
+
+    if (footnotes.isNotEmpty) {
+      // Sort footnotes by appearing order.
+      final ordinal = <String, int>{
+        for (int i = 0; i < footnoteLabels.length; i++)
+          'fn-${footnoteLabels[i]}': i,
+      };
+      footnotes.sort((l, r) {
+        final idl = l.attributes['id']?.toLowerCase() ?? '';
+        final idr = r.attributes['id']?.toLowerCase() ?? '';
+        return (ordinal[idl] ?? 0) - (ordinal[idr] ?? 0);
+      });
+      final list = Element('ol', footnotes);
+
+      // Ignore GFM attribute: <data-footnotes>.
+      final section = Element('section', [list])
+        ..attributes['class'] = 'footnotes';
+      blocks.add(section);
+    }
+    return blocks;
+  }
+
+  /// Generate backref nodes, append them to footnote definition's last child.
+  void _appendBackref(List<Node> children, String ref, int count) {
+    final refs = () sync* {
```

Comment:
Unfortunately `sync*` is slow. https://github.com/dart-lang/sdk/issues/32102

Why do we need a self-executing block here? (Apologies if I suggested it 😅 )

It looks like this can be:

```dart
var refs = [
  for (var i = 0; i < count; i++)
    ...[
      Text(' '),
      Element('a', [
          Text('\u21a9'),
          if (i > 0)
            Element('sup', [Text(num)])..attributes['class'] = 'footnote-ref',
        ])
            // Ignore GitHub's attributes: <data-footnote-backref aria-label="Back to content">.
              ..attributes['href'] = '#fnref-$ref$suffix'
              ..attributes['class'] = 'footnote-backref',
    ],
];
```

OK so that is _really_ not readable, haha, so I would extract out some logic into a either a real or a fake constructor on Element, like:

```dart
extension on Element {
  static Element footnoteAnchor(String ref, int index) {
    final number = '${i + 1}';
    final suffix = i > 0 ? '-$number' : '';
    return Element('a', [
          Text('\u21a9'),
          if (index > 0)
            Element('sup', [Text(number)])..attributes['class'] = 'footnote-ref',
        ])

          // Ignore GitHub's attributes: `<data-footnote-backref aria-label="Back to content">`.
          ..attributes['href'] = '#fnref-$ref$suffix'
          ..attributes['class'] = 'footnote-backref';
  }
}
```

Then the above list mess can become:


```dart
var refs = [
  for (var i = 0; i < count; i++)
    ...[Text(' '), Element.footnoteAnchor(ref, i)],
];
```